easybuild/framework/easyblock.py

```diff
@@ -1818,41 +1818,43 @@ def skip_extensions_parallel(self, exts_filter):
         self.log.experimental("Skipping installed extensions in parallel")
         print_msg("skipping installed extensions (in parallel)", log=self.log)
 
-        async_cmd_info_cache = {}
-        running_checks_ids = []
+        thread_pool = ThreadPoolExecutor(max_workers=self.cfg['parallel'])
+
+        async_shell_cmd_tasks = []
         installed_exts_ids = []
         exts_queue = list(enumerate(self.ext_instances[:]))
         checked_exts_cnt = 0
         exts_cnt = len(self.ext_instances)
+        done_tasks = []
 
         # asynchronously run checks to see whether extensions are already installed
-        while exts_queue or running_checks_ids:
+        while exts_queue or async_shell_cmd_tasks:
 
             # first handle completed checks
-            for idx in running_checks_ids[:]:
+            for task in done_tasks:
+                async_shell_cmd_tasks.remove(task)
+                res = task.result()
+                idx = res.task_id
                 ext_name = self.ext_instances[idx].name
-                # don't read any output, just check whether command completed
-                async_cmd_info = check_async_cmd(*async_cmd_info_cache[idx], output_read_size=0, fail_on_error=False)
-                if async_cmd_info['done']:
-                    out, ec = async_cmd_info['output'], async_cmd_info['exit_code']
-                    self.log.info("exts_filter result for %s: exit code %s; output: %s", ext_name, ec, out)
-                    running_checks_ids.remove(idx)
-                    if ec == 0:
-                        print_msg("skipping extension %s" % ext_name, log=self.log)
-                        installed_exts_ids.append(idx)
-
-                    checked_exts_cnt += 1
-                    exts_pbar_label = "skipping installed extensions "
-                    exts_pbar_label += "(%d/%d checked)" % (checked_exts_cnt, exts_cnt)
-                    self.update_exts_progress_bar(exts_pbar_label)
+                self.log.info(f"exts_filter result for {ext_name}: exit code {res.exit_code}; output: {res.output}")
+                if res.exit_code == 0:
+                    print_msg(f"skipping extension {ext_name}", log=self.log)
+                    installed_exts_ids.append(idx)
+
+                checked_exts_cnt += 1
+                exts_pbar_label = "skipping installed extensions "
+                exts_pbar_label += "(%d/%d checked)" % (checked_exts_cnt, exts_cnt)
+                self.update_exts_progress_bar(exts_pbar_label)
 
             # start additional checks asynchronously
-            while exts_queue and len(running_checks_ids) < self.cfg['parallel']:
+            while exts_queue:
```

Comment:
Is it necessary to trigger this loop for each iteration on `while exts_queue or async_shell_cmd_tasks`? This loop is just submitting the tasks for all extensions and `exts_queue` is not modified during the outer loop.

So the whole outer loop  `while exts_queue or async_shell_cmd_tasks:` could be broken down into 2 independent loops:
```
for idx, ext  in exts_queue:
    cmd, stdin = resolve_exts_filter_template(exts_filter, ext)
    task = thread_pool.submit(run_shell_cmd, cmd, ...)
    async_shell_cmd_tasks.append(task)

while async_shell_cmd_tasks:
    (done_tasks, _) = wait(async_shell_cmd_tasks, timeout=1, return_when=FIRST_COMPLETED)
    for task in done_tasks:
        async_shell_cmd_tasks.remove(task)
        res = task.result()
        [...]
        self.update_exts_progress_bar(exts_pbar_label)
```

Reply:
fixed in ca3828b